### PR TITLE
fix: (terminal) buffer previewer

### LINKED ISSUE #714: Colorize terminal buffer preview
**Describe the solution you'd like**

It would be awesome to colorize terminal buffer preview for more visibility.

Current previewer does not invoke buffer related autocommands maybe for performance perspective. Could we invoke some color related autocommands only when the buffer is terminal?

### PR DESCRIPTION
**Pull request by fdschmidt93**
Would close #714 by introducing a new previewer that reuses existing buffers.

@Conni2461 Idk, personally a bit torn on this. While `previewers.new_buffer_previewer` feels backward for `builtin.buffers` that we already have, though now ofc the buffer for `actions.delete` has to be replaced prior to deletion and the solution generally feels bloat. I'm not sure there are other side effects (that's why I abstained touching `new_buffer_previewer`, as we'd have conflicting logic ala creating buffers vs using existing ones).

I'm personally fine with terminal buffers not being highlighted ;)

### CODE REVIEW

**Comment by Conni2461** [comment]
I wouldn't touch `new_buffer_previewer` for this, i had something like this in mind when mentioning that i need to restructure it from in #714, then i ignored it for 4 and half months because i dont use previewer for `builtin.buffers` :rofl: 
Originally i was to lazy to write another base Previewer so i just used `vimgrep` and it worked till someone mentioned "special" buffers and i just introduced a silly workaround. We can remove those which probably will make PR smaller, so i dont actually have a problem doing this PR :+1: 
We can delete this https://github.com/nvim-telescope/telescope.nvim/blob/f1a27baf279976845eb43c65e99a71d7f0f92d02/lua/telescope/previewers/term_previewer.lua#L297-L300 and this https://github.com/nvim-telescope/telescope.nvim/blob/f1a27baf279976845eb43c65e99a71d7f0f92d02/lua/telescope/previewers/buffer_previewer.lua#L373-L378

File: lua/telescope/previewers/buffer_previewer.lua
Comment (Conni2461) [inline lua/telescope/previewers/buffer_previewer.lua:887]: I think you need to make sure that buffer is loaded. So `if vim.api.nvim_buf_loaded ==false then vim.fn.bufload` or so

For example if you load a session, all buffers are unloaded, so you will end up with no highlighting or something like that.

File: lua/telescope/previewers/buffer_previewer.lua
Comment (Conni2461) [inline lua/telescope/previewers/buffer_previewer.lua:907]: We should reuse this. So local function and then just setting it in `new_buffer_previewer` and here.

**Comment by Conni2461** [comment]
otherwise LGTM thanks for picking up my slack :)

File: lua/telescope/previewers/buffer_previewer.lua
Comment (Conni2461) [reply lua/telescope/previewers/buffer_previewer.lua:887]: Okay just tested we dont need to. In the past in ran into this but maybe that was bug upstrem and `nvim_win_set_buf` now does ensure that. So better ignore that comment here :rofl:

File: lua/telescope/previewers/buffer_previewer.lua
Comment (fdschmidt93) [reply lua/telescope/previewers/buffer_previewer.lua:907]: This is a lot more funny than I've thought. It has the side effect that it scrolls the actual buffer in the background when pressing <c-u> for `builtin.buffers` :laughing: I've refactored `scroll_fn` to scroll on the window as opposed to the buffer, which has _another_ funny effect. This is relatively easy to make consistently.

**Comment by fdschmidt93** [comment]
There is a crazy stupid issue with this I've so far just overlooked in my tests (please confirm if you have the time). If we use actual buffers for the preview, the fact that the popup style is `minimal` messes the buffer look once attached to the new window.

This does not happen if you select a buffer that is currently not opened in the window the picker was launched from but was opened in that window with the right settings before. Fade me.

File: lua/telescope/previewers/buffer_previewer.lua
Comment (Conni2461) [reply lua/telescope/previewers/buffer_previewer.lua]: `nvim_win_call` works for me

File: lua/telescope/builtin/internal.lua
Comment (Conni2461) [inline lua/telescope/builtin/internal.lua:661]: this will break if `previewer = false` is set. i think

And then we dont actually need any to do anything

File: lua/telescope/previewers/buffer_previewer.lua
Comment (Conni2461) [inline lua/telescope/previewers/buffer_previewer.lua:879]: We can just set `self.state.winid` in `preview_fn` every time. I dont really see a reason to give setup access to status

File: lua/telescope/previewers/buffer_previewer.lua
Comment (Conni2461) [inline lua/telescope/previewers/buffer_previewer.lua:892]: Current master sets `pcall(vim.api.nvim_buf_add_highlight, bufnr, ns_previewer, "TelescopePreviewLine", lnum - 1, 0, -1
` to indicate the current line, we should also do something similar here.

**Comment by Conni2461** [comment]
> There is a crazy stupid issue with this I've so far just overlooked in my tests (please confirm if you have the time). If we use actual buffers for the preview, the fact that the popup style is minimal messes the buffer look once attached to the new window.

Good old weird bugs, i have not missed you

File: lua/telescope/previewers/buffer_previewer.lua
Comment (fdschmidt93) [reply lua/telescope/previewers/buffer_previewer.lua:879]: I've actually just refactored, such that `setup` and `teardown` take care of setting the window options (as opposed to in `builtin.buffers`) accordingly for which I would need status to get `original_win_id`. Fine to keep status here then?

File: lua/telescope/builtin/internal.lua
Comment (fdschmidt93) [reply lua/telescope/builtin/internal.lua:661]: Yeah, I've refactored it to use `previewer.setup` and `previewer.teardown` already. See the other comment whether you are fine with this.

File: lua/telescope/previewers/buffer_previewer.lua
Comment (fdschmidt93) [reply lua/telescope/previewers/buffer_previewer.lua:892]: Agree we should, but I have to think more about how to. `nvim_buf_add_highlight` cannot be used, since it's highlighting the buffer of course in previewer and background :laughing:

File: lua/telescope/previewers/buffer_previewer.lua
Comment (Conni2461) [reply lua/telescope/previewers/buffer_previewer.lua:879]: then yeah

File: lua/telescope/previewers/buffer_previewer.lua
Comment (Conni2461) [reply lua/telescope/previewers/buffer_previewer.lua:892]: i am kinda glad i skipped this issue, seems like a lot of work for terminal highlighting

Comment (fdschmidt93) [reply lua/telescope/previewers/buffer_previewer.lua:892]: I think this is not impossible with `nvim_win_set_decoraton_provider` :laughing:

**Comment by fdschmidt93** [comment]
Actually, it's intended behavior.. Just not really nice one as it's been probably set without floats in mind.

> When editing a buffer that has been edited before, the options from the window
that was last closed are used again.  If this buffer has been edited in this
window, the values from back then are used.  Otherwise the values from the
last closed window where the buffer was edited last are used.
It's possible to set a local window option specifically for a type of buffer.
When you edit another buffer in the same window, you don't want to keep
using these local window options.  Therefore Vim keeps a global value of the
local window options, which is used when editing another buffer.  Each window
has its own copy of these values.  Thus these are local to the window, but
global to all buffers in the window. 

`:h local-options`

File: lua/telescope/previewers/buffer_previewer.lua
Comment (fdschmidt93) [reply lua/telescope/previewers/buffer_previewer.lua:945]: Yeah, I've got it working with `nvim_win_set_decoration_provider`. I require Github Sponsors from anyone who upvoted the original issue :laughing:

**Comment by fdschmidt93** [comment]
Well, that escalated quite a lot. It's done though it's quite complicated for a fix. While I cannot think of anything wrong with this right now, I'll test this more tomorrow or next week I suppose.

Complexity mostly stems from two things:

- Low-level api `vim.api.nvim_set_decoration_provider` needed to set buffer-window local extmarks -- it's unclear to me from docs whether this is adequately cleared but I think I cannot do more to assure this
- `:h local-options` -- if a buffer hasn't yet been attached to a window, it will carry over the minimal layout from the popup -- doing my best to recover from this, but maybe fillchars (which seem fine in my tests) or `EndOfBuffer` (doesn't seem to be cleared as per `:h vim.api.nvim_open_win`) are still acting up?

File: lua/telescope/builtin/internal.lua
Comment (Conni2461) [reply lua/telescope/builtin/internal.lua:647]: Yeah its still broken if `previewer = false`. 
`E5108: Error executing lua ...lugged/telescope.nvim/lua/telescope/builtin/internal.lua:647: Argument "pos" must be a [row, col] array`

We do not have `col` by default just do `entry.col or 1`.Or dont enhance the action if `opts.previewer = false`

File: lua/telescope/builtin/internal.lua
Comment (Conni2461) [inline lua/telescope/builtin/internal.lua:829]: we are missing a newline :sob:

**Review by Conni2461 — CHANGES_REQUESTED** [review]
LGTM but its still broken if `previewer = false` see one of the old conversations are opened it again.

Thanks :)

File: lua/telescope/builtin/internal.lua
Comment (fdschmidt93) [reply lua/telescope/builtin/internal.lua:829]: What have I done? :scream: and why does neither luacheck or stylua check complain (and ofc my stylua not take care of this)

File: lua/telescope/builtin/internal.lua
Comment (fdschmidt93) [reply lua/telescope/builtin/internal.lua:647]: Yeah, that was just a stupid oversight on my part, sorry; I've burnt a bit on this PR and overread in coop with the other issue :sweat_smile: 

Will fix later.

Comment (fdschmidt93) [reply lua/telescope/builtin/internal.lua]: `column` in `nvim_win_set_cursor` is zero-indexed, so `entry.col or 0`; I also mix up zero and one indexing across functions more often than I'd like :laughing:

**Comment by fdschmidt93** [comment]
Don't merge this yet.. if previewer is false, the window options are not applied :facepalm:

**Comment by Conni2461** [comment]
> Don't merge this yet.. if previewer is false, the window options are not applied facepalm

Window options are applied just fine for me and i dont even see why this should be a problem. `previewer = false` hasn't changed in this PR really besides the additional action doing the a second jump. Works in master and this PR just fine

**Comment by Conni2461** [comment]
Oh but buffer deletion is broken

**Comment by fdschmidt93** [comment]
> Window options are applied just fine for me and i dont even see why this should be a problem. previewer = false hasn't changed in this PR really besides the additional action doing the a second jump. Works in master and this PR just fine

Yes, you're right. It cannot be an issue because the preview window is never created which could overwrite the buffer-window local options. But in one state I saw the options not applied and I cannot reproduce this now :laughing: I'm too fried.

**Comment by fdschmidt93** [comment]
`action.delete_buffer` is now fine. Window options seem to be working as intended as well but I don't like having to wrap it in `builtin.buffers` itself.

**Comment by fdschmidt93** [comment]
I've also tested whether `nvim_set_decoration_provider` is cleared by `nvim_buf_clear_namespace` and it's not. So if a user would ever create a window with the same id as preview window (s)he would have the cursorline highlighted with `TelescopePreviewLine`. Travelling this weekend, so I'll look into it thereafter.

**Comment by fdschmidt93** [comment]
@Conni2461 could you please give this another whirl? `previewer=false` and `action.delete_buffer` are fixed. Sorry, couldn't test myself before you've took a crack at it :sweat_smile: 

I've asked on nvim gitter about clearing decoration provider and didn't get an answer (don't wanna ping bfredl, but I'm sure the circle of people to answer this definitively is very exclusive right now :laughing: )

I opened the buffer previewer 30 times in a session and virtually nothing happened in htop. I've minimized callbacks from `set_decoration_provider` as you'll see in [diff](https://github.com/nvim-telescope/telescope.nvim/pull/1120/commits/691be2a156d088d35baf7cf6aad8f554da7d1dc9) of final commit via returning `false` in `on_start` once previewer is closed.

**Review by Conni2461 — APPROVED** [review]
I think this is good. Thanks :)

~~And i also think this might fix https://github.com/nvim-telescope/telescope.nvim/issues/1145 which happens since today but not with this PR, no idea why or how this PR should affect this~~

**Comment by Conni2461** [comment]
Oh wait it doesn't happen here because this branch is not yet rebased. So 1145 is a actual bug we need to address. Looking into it

**Pull request merged**
